### PR TITLE
feat: added support for filtering by author

### LINKED ISSUE #2380: [feature request] add ability to filter how to / research list by created by user
**Is your feature request related to a problem? Please describe.**
No problem as such, nice to have feature or improvement [raised on discord](https://discord.com/channels/586676777334865928/1110629403857268766).

**Describe the solution you'd like**
Improvements to search feature so that it is possible to specifically search by created user.

**Describe alternatives you've considered**
Manually check list for items created by user interested in.

**Additional context**

### PR DESCRIPTION
**Pull request by goratt12**
PR Checklist

- [ ] - Commit [messages are descriptive](https://github.com/ONEARMY/community-platform/blob/master/CONTRIBUTING.md#--commit-style-guide), it will be used in our [Release Notes](https://github.com/ONEARMY/community-platform/releases/)

PR Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Developer experience (improves developer workflows for contributing to the project)

## Description

Add the ability to filter by author in HowTo and Research pages

## Git Issues

Closes #2380

## Screenshots/Videos

image
image

---

## What happens next?

Thanks for the contribution! We try to make sure all PRs are reviewed ahead of a monthly dev call (first Monday of the month, open to all!).

If the PR is working as intended it'll be merged and included in the next platform release, if not changes will be requested and re-reviewed once updated.

If you need more immediate feedback you can try reaching out on Discord in the [Community Platform `development` channel](https://discord.com/channels/586676777334865928/938781727017558018).

### CODE REVIEW

**Comment by davehakkens** [comment]
hey thanks for working on a solution @goratt12, works smooth.
However i'm not sure if this is was the intention for the original issue since it takes up quite some important space on top. 
Let me know if you have some suggestions on the original issue @iSCJT or let me know if this would be useful to release like this for Precious Plastic @thisismattia 

**Alternatively**
Would it also be possible to try and optimize the search? Searchbar:  > `Search for How-to or Username` I doubt that searching for usernames is a common thing, this would bring the function in without taking to much space.
- There is some code in place that allows us to define with the search priority (currently priority is how-to title, text and later username. We could increase priority on username, i believe this this should be easy.
- The hard part: Not sure how to fix slow loading times

**Comment by goratt12** [comment]
Hi,
Regarding the slow loading times, I suggest moving the function that call the search to `onBlur` instead of `onChange`.
this should decrease the load time, because currently the `fuse.search` (the slow function) is called every time a user is typing and it can spam the call stack.
https://github.com/ONEARMY/community-platform/blob/b73ad41ea2728b9048ae0f1d87c6a7835ed734ad/src/pages/common/SortFilterHeader/SortFilterHeader.tsx#L120-L124

**Comment by goratt12** [comment]
> hey thanks for working on a solution @goratt12, works smooth. However i'm not sure if this is was the intention for the original issue since it takes up quite some important space on top. Let me know if you have some suggestions on the original issue @iSCJT or let me know if this would be useful to release like this for Precious Plastic @thisismattia
> 
> **Alternatively** Would it also be possible to try and optimize the search? Searchbar: > `Search for How-to or Username` I doubt that searching for usernames is a common thing, this would bring the function in without taking to much space.
> 
> * There is some code in place that allows us to define with the search priority (currently priority is how-to title, text and later username. We could increase priority on username, i believe this this should be easy.
> * The hard part: Not sure how to fix slow loading times

Hey.
I tried increasing the weight of the field `_createdBy` in the search (fuze) and it didn't resolve the issue.
there are still unrelated items showing in the list after I searched for specific user.

**Comment by thisislawatts** [comment]
@davehakkens, I think the filtering by Author is a useful filter to add. Could we reduce the column width slightly to create a bit more space? The existing drop downs have a lot of unused whitespace in them due to the word length.

Top: Existing
Bottom: New dropdown with each column width reduced by ~15%
Screenshot 2023-09-30 at 21 31 50

**Comment by goratt12** [comment]
I have a suggestion, we could use a URL param state for the author filtering. example: `/research?author=goratt12`
and in the user profile page we can have link to the research/howto page with the author param so it will show the user articles

**Comment by davehakkens** [comment]
Not very convinced yet @thisislawatts, also because it brings some extra complexcity on mobile. I would prefer to bring it up in the moderators call Monday to hear their preference. 

Sounds nice @goratt12 but we currently have a smiliar feature like that in beta which i think solves the same problem?
Its shows a list of all the how-to and research directly on the user profile 
afbeelding

**Comment by goratt12** [comment]
That's cool!
we can have both.
make these clickable and redirect to the relevant page with a URL filter.
271827724-7b99fea2-bc99-4506-8be0-6b5aec1ba956

**Comment by davehakkens** [comment]
> That's cool! we can have both. make these clickable and redirect to the relevant page with a URL filter. 271827724-7b99fea2-bc99-4506-8be0-6b5aec1ba956

That seems like a nice little detail. But might be better to create a new issue/PR for that. Would you be up for creating that @goratt12? -since its falls a bit outside the scope of this one..

**Comment by davehakkens** [comment]
> @davehakkens, I think the filtering by Author is a useful filter to add. Could we reduce the column width slightly to create a bit more space? The existing drop downs have a lot of unused whitespace in them due to the word length.
> 
> Top: Existing Bottom: New dropdown with each column width reduced by ~15% Screenshot 2023-09-30 at 21 31 50

Had a talk in the teams about this. Everyone is good to merge this one it with your suggestions @thisislawatts of reducing the width

**Comment by goratt12** [comment]
> > That's cool! we can have both. make these clickable and redirect to the relevant page with a URL filter. 271827724-7b99fea2-bc99-4506-8be0-6b5aec1ba956
> 
> That seems like a nice little detail. But might be better to create a new issue/PR for that. Would you be up for creating that @goratt12? -since its falls a bit outside the scope of this one..

Yeah no problem, I'll open the issue after this PR will merge so I can base it on the author filter

File: src/stores/Howto/howto.store.tsx [hunk 1]
```diff
@@ -107,6 +111,10 @@ export class HowtoStore extends ModuleStore {
     this.activeSorter = sorter
   }
 
+  public updateSelectedAuthor(author: string) {
+    this.selectedAuthor = author
```
Comment (AlfonsoGhislieri) [inline src/stores/Howto/howto.store.tsx:115]: same suggestion as above

File: src/pages/common/SortFilterHeader/SortFilterHeader.tsx [hunk 2]
```diff
@@ -61,9 +61,17 @@ export const SortFilterHeader = ({
       : ''
 
   const [sortState, setSortState] = useState(defaultSortingOption)
+  const [selectedAuthor, setSelectedAuthor] = useState(null)
+
```
Comment (AlfonsoGhislieri) [inline src/pages/common/SortFilterHeader/SortFilterHeader.tsx:65]: **Issue**: We shouldn't need to use state for the implementation, this should be observable with `selectedAuthor` and automatically will cause the component to re-render when it is modified using the `updateSelectedAuthor` function

**Comment by AlfonsoGhislieri** [comment]
Thanks for the changes @goratt12 just some conflicts that need to be sorted out

**Comment by goratt12** [comment]
> Thanks for the changes @goratt12 just some conflicts that need to be sorted out

Done :)

**Review by AlfonsoGhislieri — APPROVED** [review]

**Pull request merged**